File: PolyScript/frameworks/test/TestCompiler.cs
/*
 * Test C# Compiler Tool for PolyScript Framework
 * CRUD × Modes Architecture: Zero-boilerplate CLI development
 * 
 * Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
 */

using System;
using System.Collections.Generic;
using PolyScript.Framework;
using Spectre.Console;

namespace PolyScript.Test
{
    [PolyScriptTool]
    [Rebadge("compile", "create+live")]
    [Rebadge("dry-compile", "create+simulate")]
    [Rebadge("status", "read+live")]
    [Rebadge("clean", "delete+live")]
    public class TestCompilerTool : IPolyScriptTool
    {
        public string Description => "Test C# compiler tool demonstrating CRUD × Modes";

        public object Create(string resource, Dictionary<string, object> options, PolyScriptContext context)
        {
            context.Log($"Compiling {resource}...");

            var outputFile = options.GetValueOrDefault("output", resource?.Replace(".cs", ".exe")) ?? "output.exe";
            var optimize = options.GetValueOrDefault("optimize", false);

            return new
            {
                compiled = resource,
                output = outputFile,
                optimized = optimize,
                timestamp = DateTime.Now.ToString("O")
            };
        }

        public object Read(string resource, Dictionary<string, object> options, PolyScriptContext context)
        {
            context.Log("Checking compilation status...");

            var sourceFiles = string.IsNullOrEmpty(resource) 
                ? new[] { "Program.cs", "Utils.cs", "Config.cs" }
                : new[] { resource };

            return new
            {
                source_files = sourceFiles,
                compiled_files = new[] { "Program.exe", "Utils.dll" },
                missing = new[] { "Config.dll" },
                last_build = DateTime.Now.ToString("O")
            };
        }

        public object Update(string resource, Dictionary<string, object> options, PolyScriptContext context)
        {
            context.Log($"Recompiling {resource}...");

            var incremental = options.GetValueOrDefault("incremental", false);

            return new
            {
                recompiled = resource,
                reason = "source file changed",
                incremental = incremental,
                timestamp = DateTime.Now.ToString("O")
            };
        }

        public object Delete(string resource, Dictionary<string, object> options, PolyScriptContext context)
        {
            context.Log($"Cleaning {resource ?? "build artifacts"}...");

            return new
            {
                cleaned = new[] { "*.exe", "*.dll", "*.pdb", "bin/", "obj/" },
                freed_space = "15.7 MB",
                timestamp = DateTime.Now.ToString("O")
            };
        }

        // Optional validation method for sandbox mode
        public Dictionary<string, string> ValidateCreate(string resource, Dictionary<string, object> options, PolyScriptContext context)
        {
            return new Dictionary<string, string>
            {
                ["compiler_available"] = "verified",
                ["source_exists"] = !string.IsNullOrEmpty(resource) ? "verified" : "failed",
                ["disk_space"] = "sufficient"
            };
        }
    }

    public class Program
    {
        public static int Main(string[] args)
        {
            return PolyScriptFramework.Run<TestCompilerTool>(args);
        }
    }
}

File: PolyScript/frameworks/test/TestCompiler.vb
'
' Test VB.NET Compiler Tool for PolyScript Framework
' CRUD × Modes Architecture: Zero-boilerplate CLI development
' 
' Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
'

Imports System
Imports System.Collections.Generic
Imports PolyScript.Framework

<PolyScriptTool>
<Rebadge("compile", "create+live")>
<Rebadge("dry-compile", "create+simulate")>
<Rebadge("status", "read+live")>
<Rebadge("clean", "delete+live")>
Public Class TestCompilerTool
    Implements IPolyScriptTool

    Public ReadOnly Property Description As String Implements IPolyScriptTool.Description
        Get
            Return "Test VB.NET compiler tool demonstrating CRUD × Modes"
        End Get
    End Property

    Public Function Create(resource As String, options As Dictionary(Of String, Object), context As PolyScriptContext) As Object Implements IPolyScriptTool.Create
        context.Log($"Compiling {resource}...")
        
        Dim outputFile = If(options.ContainsKey("output"), options("output").ToString(), resource?.Replace(".vb", ".exe"))
        Dim optimize = If(options.ContainsKey("optimize"), CBool(options("optimize")), False)
        
        Return New Dictionary(Of String, Object) From {
            {"compiled", resource},
            {"output", outputFile},
            {"optimized", optimize},
            {"timestamp", DateTime.Now.ToString("O")}
        }
    End Function

    Public Function Read(resource As String, options As Dictionary(Of String, Object), context As PolyScriptContext) As Object Implements IPolyScriptTool.Read
        context.Log("Checking compilation status...")
        
        Return New Dictionary(Of String, Object) From {
            {"source_files", New List(Of String) From {"Program.vb", "Utils.vb", "Config.vb"}},
            {"compiled_files", New List(Of String) From {"Program.dll", "Utils.dll"}},
            {"missing", New List(Of String) From {"Config.dll"}},
            {"last_build", DateTime.Now.AddHours(-2).ToString("O")}
        }
    End Function

    Public Function Update(resource As String, options As Dictionary(Of String, Object), context As PolyScriptContext) As Object Implements IPolyScriptTool.Update
        context.Log($"Recompiling {resource}...")
        
        Return New Dictionary(Of String, Object) From {
            {"recompiled", resource},
            {"reason", "source file changed"},
            {"incremental", If(options.ContainsKey("incremental"), CBool(options("incremental")), False)},
            {"timestamp", DateTime.Now.ToString("O")}
        }
    End Function

    Public Function Delete(resource As String, options As Dictionary(Of String, Object), context As PolyScriptContext) As Object Implements IPolyScriptTool.Delete
        context.Log($"Cleaning {If(resource, "build artifacts")}...")
        
        Return New Dictionary(Of String, Object) From {
            {"cleaned", New List(Of String) From {"*.exe", "*.dll", "*.pdb", "bin/", "obj/"}},
            {"freed_space", "22.3 MB"},
            {"timestamp", DateTime.Now.ToString("O")}
        }
    End Function
End Class

Module Program
    Sub Main(args As String())
        PolyScriptFramework.Run(Of TestCompilerTool)(args)
    End Sub
End Module